feat(react-native): add/improve existing deep linking guide (#796)
GetStream/stream-video-js: packages/react-native-sdk/docusaurus/docs/reactnative/06-advanced/06-deeplinking.mdx

---
title: Deep Linking
description: Tutorial to integrate deep linking for the calls
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

This guide talks about deep linking into a [video call](../02-tutorials/01-video-calling.mdx) or a generic application directly. It also focuses on getting the call id or similar entity through a Deep link URL and thereby use it to start a call.

> Our API makes it very easy to integrate and start with your custom UI, or our existing screens powered by our UI components.

Reviewer: I am not sure why this sentence is necessary in a deeplinking guide where we use no API of stream

Author: Yes, I will remove it

```

- `data` - Add one or more `<data>` tags, each of which represents a URI format that resolves to the activity. At minimum, the `<data>` tag must include the `android:scheme` attribute.

```xml
<data android:scheme="http" />
<data android:scheme="https" />
<!-- The URL here must exclude the scheme -->
<data android:host="`YOUR URL HERE`" />
```

- `category` - Include the `DEFAULT` and the `BROWSABLE` category. It is required in order for the intent filter to be accessible from a web browser. Without it, clicking a link in a browser cannot resolve to your app.

```xml
<category android:name="android.intent.category.DEFAULT" />
<category android:name="android.intent.category.BROWSABLE" />
```

Combined together your intent filter should look like:

```xml
<intent-filter android:autoVerify="true">
    <action android:name="android.intent.action.VIEW" />
    <category android:name="android.intent.category.DEFAULT" />
    <category android:name="android.intent.category.BROWSABLE" />
    <data android:scheme="http" />
    <data android:scheme="https" />
    <!-- Example: "stream-calls-dogfood.vercel.app” -->
    <data android:host="`YOUR URL HERE`" />
</intent-filter>
```

> #### Verifying the link

Reviewer: ```suggestion
#### Adding Asset Links
```
Keeps the naming similar to the other section `Adding Associated Domains Entitlement`

- Add intent filters that contain the `autoVerify` attribute(as we have done in our example above). This attribute signals to the system that it should verify whether your app belongs to the URL domains used in your intent filters.

- [Declare the association](https://developer.android.com/training/app-links/verify-android-applinks#web-assoc) between your website and your intent filters by hosting a Digital Asset Links JSON file at the following location:

```
https://domain.name/.well-known/assetlinks.json
```

You can use [this tool](https://developers.google.com/digital-asset-links/tools/generator) to generate the `assetlink.json` for your web project. Enter the Hosting site name, App package name and App package fingerprint(SHA256) to generate it.

- Hosting site name is the URL of the website which we will use for deep linking in our app.
- App package name can be easily found in the first line of your `MainActivity.java` file as a value to `package`.
- To get your App package fingerprint(SHA256) follow [this](https://developers.google.com/android/guides/client-auth) guide.

</TabItem>

<TabItem value='ios'>

> #### Adding deep linking support in `AppDelegate.m`

Reviewer: ```suggestion
#### Adding deep linking support in AppDelegate
```

Because it can be .m or .mm

For adding the support of deep linking in iOS we will have to follow the below steps:

1. We need to link `RCTLinking` to the project by following steps. This will be responsible to listen to incoming app links. For this, go to `AppDelegate.m` or `AppDelegate.mm` and add the following snippet:

```mm
#import <React/RCTLinkingManager.h>

- (BOOL)application:(UIApplication *)application
   openURL:(NSURL *)url
   options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
{
  return [RCTLinkingManager application:application openURL:url options:options];
}

> ```

Reviewer: we miss the statement for addition in `restorationHandler ` here..

```
- (BOOL)application:(UIApplication *)application continueUserActivity:(nonnull NSUserActivity *)userActivity
 restorationHandler:(nonnull void (^)(NSArray<id<UIUserActivityRestoring>> * _Nullable))restorationHandler
{
 return [RCTLinkingManager application:application
                  continueUserActivity:userActivity
                    restorationHandler:restorationHandler];
}
```

}
```

##### Adding Associated Domains Entitlement

Within your Signing & Capabilities section of Xcode, add the Associated Domains entitlement, thus giving your app the capability to support external domains for not only Universal Links, but web credentials too.

![XCode Signing Capabilities](../assets/06-advanced/06-deeplinking/xcode-signing-capabilitites.png)

To add Associated Domains support click on +(plus) button just below the `Signing & Capabilities` section near to `Capability`. Search for `Associated Domains` and drag and drop it to the background. This will add the Associated Domains tab.

Within the top empty field that presents itself after adding this entitlement, click on +(plus) icon and insert `applinks:` followed by your domain name.

_The `applinks:` prefix is a nod to Xcode that this domain is being used for Universal Linking, and is one of a few prefixes that Xcode will recognise depending on the technology you are implementing._

One you are done with this step, a `.entitlements` file will be created, inside `ios/YOUR_APP_NAME` folder.

Now we will add the AASA (Apple App Site Association) File to our web project. To do this, follow the steps below:

> - Create a `apple-app-site-association` file inside `.well-known` directory on the top level of your domain, so that the file is hosted at `https://domain.name/.well-known/assetlinks.json`.

Reviewer: ```suggestion
- Create a `apple-app-site-association` file inside `.well-known` directory on the top level of your domain, so that the file is hosted at `https://domain.name/.well-known/apple-app-site-association`.
```

:::note

> If you have a React project, the `.well-known/assetlinks.json` should go inside your `public` directory of the project.

Reviewer: ```suggestion
If you have a React project, the `.well-known/apple-app-site-association` should go inside your `public` directory of the project.
```

we are talking about apple here and not android

Reviewer: Also I dont think this note is necessary at all.. why is it specific for react.. and why is it only on iOS.. can we remove it please

Author: It is not specific for React, it just mentions that if you have a web project you should do it this way, so that you don't confuse yourself.

Reviewer: as discussed in slack, the removal is necessary here. 

https://getstream.slack.com/archives/C02EK5GMU8Z/p1689671683063959?thread_ts=1689667240.871759&cid=C02EK5GMU8Z

```tsx title="App.tsx"
const App = () => {
  useEffect(() => {
    const parseAndSetCallID = (url: string | null) => {
      const matchResponse = url?.match(`YOUR REGEX HERE`); // To match the paths and handle them accordingly
      if (matchResponse?.length) {
        // Your custom setup here.
      }
    };
    const { remove } = Linking.addEventListener('url', ({ url }) => {
      parseAndSetCallID(url);
    });
    const configure = async () => {
      const url = await Linking.getInitialURL();
      parseAndSetCallID(url);
    };
    configure();
    return remove;
  }, []);
};
```

You can extract the call ID if its in the URL, start the call using the call ID and navigate to appropriate screen once you have the URL as shown in the example above.

### Recap

Please do let us know if you ran into any issues while integrating deep linking using this guide in a video calling app with React Native. Our team is also happy to review your UI designs and offer recommendations on how to achieve it with Stream.

To recap what we've learned about Deep Linking:

> - The Native Setup is an important step to achieve deep linking in both Android and iOS.
> - The Native steps should be followed one after the other, sequentially.
> - To handle the Links in a React Native app the [Linking](https://reactnative.dev/docs/linking) API must be used.
> 

Reviewer: ```suggestion
- We added lines in AndroidManifest and AppDelegate for the app to be able to read the link URLs.
- We added the `apple-app-site-association` file and `assetlinks.json` file in our domain to relate our app with the domain.
- We handle the links using the built-in [Linking](https://reactnative.dev/docs/linking) API.

```

Reviewer: It's helpful to summarise the steps like the above.
